File: src/main/java/com/leijendary/spring/microservicetemplate/data/response/DataResponse.java
package com.leijendary.spring.microservicetemplate.data.response;

import com.leijendary.spring.microservicetemplate.data.PageMeta;
import com.leijendary.spring.microservicetemplate.util.RequestContextUtil;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.web.util.UriComponentsBuilder;

import java.net.URI;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

import static org.springframework.http.HttpStatus.OK;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class DataResponse<T> {

    private T data;
    private Map<String, Object> meta;
    private Map<String, URI> links;

    public static <T> DataResponseBuilder<T> builder() {
        return new DataResponseBuilder<T>()
                .status(OK)
                .selfLink();
    }

    public static class DataResponseBuilder<T> {

        private T data;
        private final Map<String, Object> meta = new HashMap<>();
        private final Map<String, URI> links = new HashMap<>();

        public DataResponse<T> build() {
            this.meta.put("timestamp", Instant.now());

            if (!this.meta.containsKey("type")) {
                throw new IllegalArgumentException("Type is not indicated");
            }

            if (!this.meta.containsKey("object")) {
                throw new IllegalArgumentException("Object is not indicated");
            }

            return new DataResponse<>(data, meta, links);
        }

        public DataResponseBuilder<T> data(final T data) {
            this.data = data;

            if (data instanceof Iterable) {
                this.meta.put("type", "array");
            } else {
                this.meta.put("type", "object");
            }

            return this;
        }

        public DataResponseBuilder<T> object(final String object) {
            this.meta.put("object", object);

            return this;
        }

        public DataResponseBuilder<T> meta(final String key, final Object value) {
            this.meta.put(key, value);

            return this;
        }

        public DataResponseBuilder<T> status(final HttpStatus httpStatus) {
            this.meta.put("status", httpStatus.value());

            return this;
        }

        public DataResponseBuilder<T> meta(final Page<?> page) {
            this.meta.put("page", new PageMeta(page));

            return this;
        }

        public DataResponseBuilder<T> selfLink() {
            this.links.put("self", RequestContextUtil.uri());

            return this;
        }

        public DataResponseBuilder<T> links(final Page<?> page) {
            final var size = page.getSize();
            final var sort = page.getSort();

            this.links.put("self", createLink(page.getPageable().getPageNumber(), size, sort));

            if (page.hasPrevious()) {
                this.links.put("previous", createLink(page.previousOrFirstPageable().getPageNumber(), size, sort));
            }

            if (page.hasNext()) {
                this.links.put("next", createLink(page.nextOrLastPageable().getPageNumber(), size, sort));
            }

            this.links.put("last", createLink(page.getTotalPages() - 1, size, sort));

            return this;
        }

        private URI createLink(final int page, final int size, final Sort sort) {
            final var uri = UriComponentsBuilder.fromUri(RequestContextUtil.uri());
            uri.replaceQueryParam("page", page + 1);
            uri.replaceQueryParam("size", size);

            if (sort.isSorted()) {
                uri.replaceQueryParam("sort", sort.toString());
            }

            return uri.build().toUri();
        }
    }
}

File: src/main/kotlin/com/leijendary/spring/template/core/data/DataResponse.kt
package com.leijendary.spring.template.core.data

import com.leijendary.spring.template.core.util.RequestContext.now
import com.leijendary.spring.template.core.util.RequestContext.uri
import io.opentelemetry.api.trace.Span
import org.springframework.data.domain.Page
import org.springframework.data.domain.Sort
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.OK
import org.springframework.web.util.UriComponentsBuilder
import java.net.URI

class DataResponse<T>(
    val data: T? = null,
    val meta: Map<String, Any> = emptyMap(),
    val links: Map<String, URI?>? = null
) : Response {
    companion object {
        fun <T> builder(): DataResponseBuilder<T> {
            return DataResponseBuilder<T>()
                .status(OK)
                .selfLink()
                .traceId()
        }
    }

    class DataResponseBuilder<T> {
        private var data: T? = null
        private val meta: MutableMap<String, Any> = HashMap()
        private val links: MutableMap<String, URI?> = HashMap()

        fun build(): DataResponse<T> {
            meta["timestamp"] = now

            return DataResponse(data, meta, links)
        }

        fun data(data: T): DataResponseBuilder<T> {
            this.data = data

            return this
        }

        fun meta(key: String, value: Any): DataResponseBuilder<T> {
            meta[key] = value

            return this
        }

        fun status(httpStatus: HttpStatus): DataResponseBuilder<T> {
            meta["status"] = httpStatus.value()

            return this
        }

        fun meta(page: Page<*>): DataResponseBuilder<T> {
            meta["page"] = PageMeta(page)

            return this
        }

        fun meta(seek: Seek<*>): DataResponseBuilder<T> {
            meta["seek"] = SeekMeta(seek)

            return this
        }

        fun traceId(): DataResponseBuilder<T> {
            meta["traceId"] = Span.current().spanContext.traceId

            return this
        }

        fun selfLink(): DataResponseBuilder<T> {
            links["self"] = uri

            return this
        }

        fun links(page: Page<*>): DataResponseBuilder<T> {
            val size = page.size
            val sort = page.sort

            links["self"] = createLink(page.pageable.pageNumber, size, sort)

            if (page.hasPrevious()) {
                val previousPageable = page.previousOrFirstPageable()

                links["previous"] = createLink(previousPageable.pageNumber, size, sort)
            }

            if (page.hasNext()) {
                val nextPageable = page.nextOrLastPageable()

                links["next"] = createLink(nextPageable.pageNumber, size, sort)
            }

            links["last"] = createLink(page.totalPages - 1, size, sort)

            return this
        }

        fun links(seek: Seek<*>): DataResponseBuilder<T> {
            val nextToken = seek.nextToken
            val limit = seek.seekable.limit

            links["next"] = createLink(nextToken, limit)

            return this
        }

        private fun createLink(page: Int, size: Int, sort: Sort): URI? {
            val path: URI = uri ?: return null
            val builder = UriComponentsBuilder.fromUri(path)
                .replaceQueryParam("page", page)
                .replaceQueryParam("size", size)

            if (sort.isSorted) {
                val sortString = sort
                    .toSet()
                    .map { "${it.property},${it.direction}" }

                builder.replaceQueryParam("sort", sortString)
            }

            return builder.build().toUri()
        }

        private fun createLink(nextToken: String?, limit: Int): URI? {
            val path = uri ?: return null
            val builder = UriComponentsBuilder.fromUri(path)
                .replaceQueryParam("limit", limit)
                .replaceQueryParam("nextToken", nextToken)

            return builder.build().toUri()
        }
    }
}